flowey/flowey_lib_common/src/publish_test_results.rs

```diff
@@ -0,0 +1,232 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+//! Publish test results.
+//!
+//! - On ADO, this will hook into the backend's native JUnit handling.
+//! - On Github, this will publish artifacts containing the raw JUnit XML file
+//!   and any optional attachments.
+//! - When running locally, this will optionally copy the XML files and any
+//!   attachments to the provided artifact directory.
+
+use crate::_util::copy_dir_all;
+use flowey::node::prelude::*;
+use std::collections::BTreeMap;
+
+flowey_request! {
+    pub struct Request {
+        /// Path to a junit.xml file
+        ///
+        /// HACK: this is an optional since `flowey` doesn't (yet?) have any way
+        /// to perform conditional-requests, and there are instances where nodes
+        /// will only conditionally output JUnit XML.
+        ///
+        /// To keep making forward progress, I've tweaked this node to accept an
+        /// optional... but this ain't great.
+        pub junit_xml: ReadVar<Option<PathBuf>>,
+        /// Brief string used when publishing the test.
+        pub test_label: String,
+        /// Additional files or directories to upload
+        /// (not used on ADO unless the associated boolean is true).
+        ///
+        /// This is a workaround for platforms that do not have JUnit integration.
+        /// Any attachments listed in the JUnit file (that would have been
+        /// automatically uploaded on ADO) should be listed here.
+        pub attachments: BTreeMap<String, (ReadVar<PathBuf>, bool)>,
+        /// Copy the xml file and attachments to the provided directory.
+        /// Only supported on local backend.
+        pub output_dir: Option<ReadVar<PathBuf>>,
+        /// Side-effect confirming that the publish has succeeded
+        pub done: WriteVar<SideEffect>,
+    }
+}
+
+new_flow_node!(struct Node);
+
+impl FlowNode for Node {
+    type Request = Request;
+
+    fn imports(ctx: &mut ImportCtx<'_>) {
+        ctx.import::<crate::ado_task_publish_test_results::Node>();
+    }
+
+    fn emit(requests: Vec<Self::Request>, ctx: &mut NodeCtx<'_>) -> anyhow::Result<()> {
+        let mut use_side_effects = Vec::new();
+        let mut resolve_side_effects = Vec::new();
+
+        let attempt = match ctx.backend() {
```

Comment:
You're doing some spooky stuff with string interpolation here that makes me worried. i.e: by having this match statement "far away" from the backend-specific "FFI" code, you're stretching the scope of those backend-specific semantics across a pretty large surface area. In the future, someone might want end up misundersatnding how things work, and - for example - writing a intermediate Rust step that tried to `println!` the `attempt` var (or whatever), but get surprised when it doesn't work.

All that is to say - I think you should make `attempt` have the type `ReadVar<String>`, which is populated by its own dedicated bit of backend-specific code. This might require defining some new constants + extending APIs like `get_ado_variable` / `get_gh_context_var`.

Yes, this will be slightly higher overhead at runtime due to the back-and-forth from YAML to rust and vice-versa... but it'll be relatively low-overhead, and result in code that is more type-safe and maintainable.

Reply:
Decided we don't need this on github, and to just put it at the step definition for ado.